tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -243,76 +248,279 @@ struct FileContent {
   }
 };
 
-class ASTProducer : public ThreadSafeRefCountedBase<ASTProducer> {
-  SwiftInvocationRef InvokRef;
-  SmallVector<BufferStamp, 8> Stamps;
-  ThreadSafeRefCntPtr<ASTUnit> AST;
-  SmallVector<std::pair<std::string, BufferStamp>, 8> DependencyStamps;
-
-  struct QueuedConsumer {
-    SwiftASTConsumerRef consumer;
-    std::vector<ImmutableTextSnapshotRef> snapshots;
-    const void *oncePerASTToken;
+/// An \c ASTBuildOperations builds an AST. Once the AST is built, it informs
+/// a list of \c SwiftASTConsumers about the built AST.
+/// It also supports cancellation with the following paradigm: If an \c
+/// SwiftASTConsumer is no longer needed, it can be cancelled, which will remove
+/// it from the \c ASTBuildOperation. If the \c ASTBuildOperation has no more
+/// consumers attached to it, it will cancel the AST build at the next
+/// opportunity.
+class ASTBuildOperation
+    : public std::enable_shared_from_this<ASTBuildOperation> {
+  /// After the AST has been built, the corresponding result.
+  struct ASTBuildResult {
+    /// The AST that was created by the build operation.
+    ASTUnitRef AST;
+    /// An error message emitted by the creation of the AST. There might still
+    /// be an AST if an error occurred, but it's usefulnes depends on the
+    /// severity of the error.
+    std::string Error;
+    /// Whether the build operation was cancelled. There might be an AST and
+    /// error but their usefulness depends on when the operation was cancelled.
+    bool Cancelled;
+    /// Whether the result contains any values, i.e. whether the operation has
+    /// produced a result yet.
+    bool HasValue;
+
+    ASTBuildResult() : HasValue(false) {}
+
+    void emplace(ASTUnitRef AST, std::string Error, bool Cancelled) {
+      assert(!HasValue && "Should only emplace a result once");
+      this->HasValue = true;
+      this->AST = AST;
+      this->Error = Error;
+      this->Cancelled = Cancelled;
+    }
+
+    operator bool() const { return HasValue; }
+
+    size_t getMemoryCost() {
+      size_t Cost = sizeof(*this) + Error.size();
+      if (AST && AST->getCompilerInstance().hasASTContext()) {
+        Cost += AST->Impl.CompInst.getASTContext().getTotalMemory();
+      } else {
+        Cost += sizeof(*AST);
```

Comment:
Shouldn't this always be part of the cost even if there is an ASTContext?

Reply:
I was just copying (and adjusting) the old implementation [here](https://github.com/apple/swift/blob/81f2be251b2375c0d47409457f490e47e112e6b8/tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp#L288-L293) which was also computing the size like this. But yes, I think we should also be including it. 

More importantly, we were derefencing `AST` in the `else` branch if `AST` was `nullptr`, which I fixed now.